File: 2D_Spiral_Array/spiral_array.cpp
#include <iostream>
#include <vector>

using namespace std;

bool isInvalid(vector<vector<int>>& res, int r, int c) {
  if (r < 0 || c < 0 || r >= (int)res.size() || c >= (int)res.size() || res[r][c] != 0) {
      return true;
  }
  return false;
}
vector<vector<int>> getSpiral(int n) {
  if (n <= 0) return {{}};

  vector<int> v(n, 0);
  vector<vector<int>> res(n, v); //(n, vector<int>(n, 0));
  vector<int> dc = {1, 0, -1, 0};
  vector<int> dr = {0, 1, 0, -1};
  int dir = 0, val = 0, r = 0, c = 0, limit = n * n;
  while (val++ < limit) {
      res[r][c] = val;
      r += dr[dir];
      c += dc[dir];
      if (isInvalid(res, r, c)) {
        r -= dr[dir];
        c -= dc[dir];
        dir = (dir + 1) % 4;
        r += dr[dir];
        c += dc[dir];
      }
   }
   return res;
}

// To execute C++, please define "int main()"
int main() {
  auto vec = getSpiral(8);
  for (const auto& v : vec) {
    for (const auto &n : v) {
      cout << n << "\t";
    }
    cout << "\n";
  }
  return 0;
}

File: 2D_Spiral_Array/spiral_array.rs
fn is_invalid(res: &Vec<Vec<i32>>, r: i32, c: i32) -> bool {
    r < 0 || c < 0 || r >= res.len() as i32 || c >= res.len() as i32 || res[r as usize][c as usize] != 0
}

fn get_spiral(n: i32) -> Vec<Vec<i32>> {
    if n <= 0 {
        return vec![vec![]];
    }

    let mut res = vec![vec![0; n as usize]; n as usize];
    let dc = vec![1, 0, -1, 0];
    let dr = vec![0, 1, 0, -1];
    let mut dir = 0;
    let mut val = 0;
    let mut r   = 0;
    let mut c   = 0;
    let limit = n * n;

    while val < limit {
        val += 1;
        res[r as usize][c as usize] = val;
        r += dr[dir];
        c += dc[dir];
        if is_invalid(&res, r, c) {
            r -= dr[dir];
            c -= dc[dir];
            dir = (dir + 1) % 4;
            r += dr[dir];
            c += dc[dir];
        }
    }
    res
}

fn main() {
    let vecs = get_spiral(8);

    for vec in vecs {
        for num in vec {
            print!("{}\t", num);
        }
        println!();
    }
}
